### PR TITLE
Support keystone with certificate

### LINKED ISSUE #209: Support keystone endpoint with certificate
Currently cluster-api-provider-openstack can not use keystone endpoint with certificate.

Python-openstackclient can use keystone endpoint with certificate by the following clouds.yaml using cacert key.

```yaml
clouds:
  openstack:
    auth:
      auth_url: https://yourauthurl:5000/v3
      username: foo
      password: bar
      project_id: foobar123
      project_name: foobar
      user_domain_name: "Default"
    cacert: /path/to/cacertfile
    region_name: "Region_1"
    interface: "public"
    identity_api_version: 3
```

Let's make cluster-api-provider-openstack to support keystone endpoint with certificate as well.

### PR DESCRIPTION
**Pull request by hidekazuna**
**What this PR does / why we need it**:

This PR supports keystone endpoint with certificate by the following clouds.yaml with cacert key.

```yaml
clouds:
  openstack:
    auth:
      auth_url: https://yourauthurl:5000/v3
      username: foo
      password: bar
      project_id: foobar123
      project_name: foobar
      user_domain_name: "Default"
    cacert: /path/to/cacertfile
    region_name: "Region_1"
    interface: "public"
    identity_api_version: 3
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #209 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

File: pkg/cloud/openstack/clients/machineservice.go [hunk 1]
```diff
@@ -95,6 +100,35 @@ func GetCloudFromSecret(kubeClient kubernetes.Interface, namespace string, secre
...
 }
 
+func authenticatedClient(ao *gophercloud.AuthOptions) (*gophercloud.ProviderClient, error) {
+
+        client, err := openstack.NewClient(ao.IdentityEndpoint)
+        if err != nil {
+                return nil, err
+        }
+        // add HTTPClient includes transport
+	caFile := "/etc//certs/cacert"
```
Comment (jichenjc) [inline pkg/cloud/openstack/clients/machineservice.go]: nit: 2 /

File: pkg/cloud/openstack/clients/machineservice.go [hunk 2]
```diff
@@ -95,6 +100,35 @@ func GetCloudFromSecret(kubeClient kubernetes.Interface, namespace string, secre
...
+	caCert, err := ioutil.ReadFile(caFile)
+	if err != nil {
+		panic(err)
```
Comment (jichenjc) [inline pkg/cloud/openstack/clients/machineservice.go]: no ca file lead to panic?

File: pkg/cloud/openstack/clients/machineservice.go [hunk 3]
```diff
@@ -95,6 +100,35 @@ func GetCloudFromSecret(kubeClient kubernetes.Interface, namespace string, secre
...
+	caCertPool.AppendCertsFromPEM(caCert)
+
+	tlsConfig := &tls.Config{
+		RootCAs: caCertPool,
+	}
+	tlsConfig.BuildNameToCertificate()
+	transport := &http.Transport{TLSClientConfig: tlsConfig}
+	client.HTTPClient = http.Client{Transport: transport}
+
+	err = openstack.Authenticate(client, *ao)
```
Comment (jichenjc) [inline pkg/cloud/openstack/clients/machineservice.go]: forgive me if I misunderstand something here.. I believe https need ca and http doesn't need it
from context we are requesting that to be mandatory https ,maybe I missed something?

File: pkg/cloud/openstack/clients/machineservice.go (hunk 1 above)
Comment (hidekazuna) [reply pkg/cloud/openstack/clients/machineservice.go]: Thanks. 
I am sorry I forgot to push so that code was old and lack of some files.

File: pkg/cloud/openstack/clients/machineservice.go (hunk 2 above)
Comment (hidekazuna) [reply pkg/cloud/openstack/clients/machineservice.go]: Thanks.
I forgot to remove authenticatedClient function which is not used anymore..

File: pkg/cloud/openstack/clients/machineservice.go (hunk 3 above)
Comment (hidekazuna) [reply pkg/cloud/openstack/clients/machineservice.go]: Yes, you are right. I should have update code https is not mandatory. I will fix.

File: cmd/clusterctl/examples/openstack/ubuntu/provider-components.yaml.template [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  userData: $WORKER_USER_DATA
+---
+apiVersion: v1
+kind: Secret
+type: Opaque
+metadata:
+  name: master-user-data
+  namespace: openstack-provider-system
+data:
+  userData: $MASTER_USER_DATA
```
Comment (jichenjc) [inline cmd/clusterctl/examples/openstack/ubuntu/provider-components.yaml.template]: this file doesn't needed now?

File: cmd/clusterctl/examples/openstack/generate-yaml.sh [hunk 5]
```diff
@@ -143,7 +144,7 @@ PASSWORD=$(echo "$OPENSTACK_CLOUD_CONFIG_PLAIN" | yq r - clouds.$CLOUD.auth.pass
 REGION=$(echo "$OPENSTACK_CLOUD_CONFIG_PLAIN" | yq r - clouds.$CLOUD.region_name)
 PROJECT_ID=$(echo "$OPENSTACK_CLOUD_CONFIG_PLAIN" | yq r - clouds.$CLOUD.auth.project_id)
 DOMAIN_NAME=$(echo "$OPENSTACK_CLOUD_CONFIG_PLAIN" | yq r - clouds.$CLOUD.auth.user_domain_name)
-
+CACERT_ORIGINAL=$(echo "$OPENSTACK_CLOUD_CONFIG_PLAIN" | yq r - clouds.$CLOUD.cacert)
```
Comment (jichenjc) [inline cmd/clusterctl/examples/openstack/generate-yaml.sh]: this should be problem if cacert doesn't exist...

File: pkg/cloud/openstack/clients/machineservice.go [hunk 6]
```diff
@@ -162,7 +167,24 @@ func NewInstanceServiceFromCloud(cloud clientconfig.Cloud) (*InstanceService, er
...
 
-	provider, err := openstack.AuthenticatedClient(*opts)
+	provider, err := openstack.NewClient(opts.IdentityEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	cloudFromYaml, err := clientconfig.GetCloudFromYAML(clientOpts)
+	if cloudFromYaml.CACertFile != "" {
+		roots, err := certutil.NewPool(caFile)
```
Comment (chrigl) [inline pkg/cloud/openstack/clients/machineservice.go]: I am thinking about the mounted caFile here. In theory, the entire code allows us to use a separate `clouds.yaml` per machine. But it is only possible to pass in a caFile for all machines.

How about extending the `cloud-config` secret with another field for the contents of the caFile.... and create a separate `x509.CertPool` from this secret?

I think it would be ok to extend the `cloud-config` secret, because it relates directly to the `clouds.yaml`, rather than having a separate secret.

File: pkg/cloud/openstack/clients/machineservice.go [hunk 7]
```diff
@@ -51,6 +55,7 @@ import (
 
 const (
 	CloudsSecretKey = "clouds.yaml"
+	caFile          = "/etc/certs/cacert"
```
Comment (chrigl) [inline pkg/cloud/openstack/clients/machineservice.go]: when extending `cloud-config`, this would go away.

**Comment by gyliu513** [comment]
@hidekazuna  can you please rebase, I want to do some test with your PR. Thanks!

**Comment by hidekazuna** [comment]
> @hidekazuna can you please rebase, I want to do some test with your PR. Thanks!

I rebased. But I will fix the code in accordance with the review comments.

File: cmd/clusterctl/examples/openstack/ubuntu/provider-components.yaml.template (hunk 4 above)
Comment (hidekazuna) [reply cmd/clusterctl/examples/openstack/ubuntu/provider-components.yaml.template]: Thanks, deleted.

**Comment by jichenjc** [comment]
thank you, I also need this as well since I moved to a https based test env... thanks ~

File: pkg/cloud/openstack/clients/machineservice.go (hunk 6 above)
Comment (jichenjc) [reply pkg/cloud/openstack/clients/machineservice.go]: @hidekazuna @gyliu513 provided this to me 
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/compare/master...cdc1807:cacert-auth

I think we can try to refer to this one, at least it works for me now on a https openstack and I think we can use secret to replace the CACERT in the commit diff

File: pkg/cloud/openstack/clients/machineservice.go (hunk 7 above)
Comment (hidekazuna) [reply pkg/cloud/openstack/clients/machineservice.go]: Instead of caFile, I added CaSecretKey like this:
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/commit/672c5affd15e14b1fb829e620087a8fe8c5c20de#diff-6f1595bb05179a80513b3baba49fbaf1R58

File: pkg/cloud/openstack/clients/machineservice.go (hunk 6 above)
Comment (hidekazuna) [reply pkg/cloud/openstack/clients/machineservice.go]: @jichenjc Thanks advice.
I updated like this: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/commit/672c5affd15e14b1fb829e620087a8fe8c5c20de#diff-6f1595bb05179a80513b3baba49fbaf1R58
I noticed I need to update getNetworkClient function in pkg/cloud/openstack/cluster/actuator.go.
But I wonder how  to fix this function as well. Could you advice?

File: pkg/cloud/openstack/clients/machineservice.go (hunk 3 above)
Comment (hidekazuna) [reply pkg/cloud/openstack/clients/machineservice.go]: Fixed in the following commit: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/commit/192e70cae1d886113b023b7b763de77e6d8c330a

File: cmd/clusterctl/examples/openstack/generate-yaml.sh (hunk 5 above)
Comment (hidekazuna) [reply cmd/clusterctl/examples/openstack/generate-yaml.sh]: Fixed in the following commit: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/commit/192e70cae1d886113b023b7b763de77e6d8c330a

File: pkg/cloud/openstack/clients/machineservice.go (hunk 6 above)
Comment (jichenjc) [reply pkg/cloud/openstack/clients/machineservice.go]: I might need more time to figure out but FYI https://github.com/kubernetes/cloud-provider-openstack has a set of https handling including network, maybe we can refer to their implementation @hidekazuna

**Comment by jichenjc** [comment]
@hidekazuna please help to rebase...

File: pkg/cloud/openstack/clients/machineservice.go (hunk 6 above)
Comment (hidekazuna) [reply pkg/cloud/openstack/clients/machineservice.go]: @chrigl @jichenjc getNetworkClient is the method of Actuator in cluster package and it's augment is *clusterv1.Cluster. We can not secrets from Actuator nor clusterv1.Cluster. I thinks this is from design issue.
Since we have to create a CA file for using OpenStack Cloud provider anyway, can't we make a compromise to use CA file?

Comment (jichenjc) [reply pkg/cloud/openstack/clients/machineservice.go]: I am not aware clusterv1 can't use secrets from Actuator, ... but conceptually yes, if you want to sync cluster to create network, apparently we need CA in order to connect to openstack, so I think it's mandatory to use CA in actuator, can we define in openstackclusterspec for this ? @hidekazuna

**Comment by gyliu513** [comment]
@morvencao @hchenxa can you help review this as well? Hope we can get this merged soon.

/cc @jichenjc @chrigl

File: cmd/clusterctl/examples/openstack/generate-yaml.sh [hunk 8]
```diff
@@ -164,11 +165,23 @@ tenant-id=\"$PROJECT_ID\"
...
+  OPENSTACK_CLOUD_PROVIDER_CONF_PLAIN="$OPENSTACK_CLOUD_PROVIDER_CONF_PLAIN
+  ca-file=\"$CACERT\"
+  "
+fi
+
 OS=$(uname)
 if [[ "$OS" =~ "Linux" ]]; then
   OPENSTACK_CLOUD_PROVIDER_CONF=$(echo "$OPENSTACK_CLOUD_PROVIDER_CONF_PLAIN"|base64 -w0)
+  if [ "$CACERT_ORIGINAL" != "null" ]; then
+    OPENSTACK_CLOUD_CACERT_CONFIG=$(cat "$CACERT_ORIGINAL"|base64 -w0)
```
Comment (hchenxa) [inline cmd/clusterctl/examples/openstack/generate-yaml.sh]: the $CACERT_ORIGINAL was the value of cacert in clouds.yaml I think, so may be we should use `echo` but not `cat` here.

refer the test result below:
```
[root@hchenxa1 openstack]#
[root@hchenxa1 openstack]# OPENSTACK_CLOUD_CONFIG_PLAIN=$(cat "clouds.yaml")
[root@hchenxa1 openstack]# CLOUD=openstack
[root@hchenxa1 openstack]# CACERT_ORIGINAL=$(echo "$OPENSTACK_CLOUD_CONFIG_PLAIN" | yq r - clouds.$CLOUD.cacert)
[root@hchenxa1 openstack]# echo $CACERT_ORIGINAL
-----BEGIN CERTIFICATE----- MIID5TCCAs2gAwIBAgIBFDANBgkqhkiG9w0BAQsFADBiMQswCQYDVQQGEwJVUzE0 MDIGA1UEChMrSW50ZXJuYXRpb25hbCBCdXNpbmVzcyBNYWNoaW5lcyBDb3Jwb3Jh dGlvbjEdMBsGA1UEAxMUSUJNIEludGVybmFsIFJvb3QgQ0EwHhcNMTYwMjI0MDUw MDAwWhcNMzUwMTAzMDQ1OTU5WjBiMQswCQYDVQQGEwJVUzE0MDIGA1UEChMrSW50 ZXJuYXRpb25hbCBCdXNpbmVzcyBNYWNoaW5lcyBDb3Jwb3JhdGlvbjEdMBsGA1UE AxMUSUJNIEludGVybmFsIFJvb3QgQ0EwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAw ggEKAoIBAQDUKGuk9Tmri43R3SauS7gY9rQ9DXvRwklnbW+3Ts8/Meb4MPPxezdE cqVJtHVc3kinDpzVMeKJXlB8CABBpxMBSLApmIQywEKoVd0H0w62Yc3rYuhv03iY y6OozBV0BL6tzZE0UbvtLGuAQXMZ7ehzxqIta85JjfFN86AO2u7xrNF0FYyGH+E0 Rn6yNhb25VrqxE0OYbSMIGoWdvS11K4SgVDqrJ9OqIk8NHrIJ8Ed24P/YPMeAp3j U409Gev1zGcuLdRr09WckQ145FZVDbPq42gcl7qYICPhZ4/eDUUjFgxpipfMGkMb 1X+Y3kFDgb4BO8Xrdda2VQo1iDZs8A8bAgMBAAGjgaUwgaIwPwYJYIZIAYb4QgEN BDIWMEdlbmVyYXRlZCBieSB0aGUgU2VjdXJpdHkgU2VydmVyIGZvciB6L09TIChS QUNGKTAOBgNVHQ8BAf8EBAMCAQYwDwYDVR0TAQH/BAUwAwEB/zAdBgNVHQ4EFgQU +d4Y5Z4wE2lRp/15hUiMfA5v2OMwHwYDVR0jBBgwFoAU+d4Y5Z4wE2lRp/15hUiM fA5v2OMwDQYJKoZIhvcNAQELBQADggEBAH87Ms8yFyAb9nXesaKjTHksLi1VKe2i zESWozYFXnRtOgOW7/0xXcfK+7PW6xwcOqvTk61fqTGxj+iRyZf2e3FNtIB+T/Lg 3SZF9sztPM0jEUELWycC8l6WPTvzQjZZBCsF+cWbU1nxvRNQluzCsTDUEIfThJIF cLu0WkoQclUrC3d2tM8jclLPssb6/OV8GaJ+4mx4ri7HbGaUAOtA/TXKR6AuhgkR NPKYhpPU0q/PRlGXdwJP8zXb8+CXMMTnI5Upur7Tc5T3I/x1Gqfz7n1sTRZfsuiQ J5uua4hz4te3oV2tm7LWcNItHD43zttBTTx/m5icg71JE2gcr2oincw= -----END CERTIFICATE-----
[root@hchenxa1 openstack]# cat $CACERT_ORIGINAL
cat: unrecognized option '-----BEGIN'
Try 'cat --help' for more information.
```

Comment (gyliu513) [reply cmd/clusterctl/examples/openstack/generate-yaml.sh]: @hchenxa Besides this issue, does there are any other issues that we need to raise with this PR?

@hidekazuna can you help update?

Comment (jichenjc) [reply cmd/clusterctl/examples/openstack/generate-yaml.sh]: I am ok now, I prefer to do some offline test before merge (http and https) ... thoughts?

Comment (gyliu513) [reply cmd/clusterctl/examples/openstack/generate-yaml.sh]: Yes, I think @hidekazuna still need some update for his PR.

File: pkg/cloud/openstack/clients/machineservice.go (hunk 6 above)
Comment (hidekazuna) [reply pkg/cloud/openstack/clients/machineservice.go]: @jichenjc Thanks, now I am trying to update.

File: cmd/clusterctl/examples/openstack/generate-yaml.sh (hunk 8 above)
Comment (hidekazuna) [reply cmd/clusterctl/examples/openstack/generate-yaml.sh]: `$CACERT_ORIGINAL` is the path of CA file. Could you provide your executed command? I tested the following.
```
./generate-yaml.sh --provider-os ubuntu -c /home/ubuntu/.config/openstack/clouds.yaml
```

Comment (hchenxa) [reply cmd/clusterctl/examples/openstack/generate-yaml.sh]: The commands example should like below:
```
[root@hchenxa1 openstack]# ./generate-yaml.sh  -h
generate-yaml.sh - generates a provider-configs.yaml file

Usage:
generate-yaml.sh [options] <path/to/clouds.yaml> <cloud> <provider os: [centos,ubuntu,coreos]>
options:
-h, --help                    show brief help
-f, --force-overwrite         if file to be generated already exists, force script to overwrite it
```

and I run the command like below:
```
[root@hchenxa1 openstack]# ./generate-yaml.sh clouds.yaml openstack centos
```

Comment (gyliu513) [reply cmd/clusterctl/examples/openstack/generate-yaml.sh]: @hchenxa can you help try

```
./generate-yaml.sh -f ./clouds.yaml openstack ubuntu
```

Comment (hidekazuna) [reply cmd/clusterctl/examples/openstack/generate-yaml.sh]: I found missed cmd/clusterctl/examples/openstack/provider-component/user-data/centos/templates/master-user-data.sh and worker-user-data.sh. I will update.

File: pkg/cloud/openstack/clients/machineservice.go (hunk 6 above)
Comment (hidekazuna) [reply pkg/cloud/openstack/clients/machineservice.go]: @jichenjc I updated by https://github.com/kubernetes-sigs/cluster-api-provider-openstack/commit/0bf8a1bf20fa24ef22c193d48b2f5bf292527979

File: cmd/clusterctl/examples/openstack/generate-yaml.sh (hunk 8 above)
Comment (hidekazuna) [reply cmd/clusterctl/examples/openstack/generate-yaml.sh]: I updated by https://github.com/kubernetes-sigs/cluster-api-provider-openstack/commit/0bf8a1bf20fa24ef22c193d48b2f5bf292527979

**Comment by hidekazuna** [comment]
Now ready for review!

File: cmd/clusterctl/examples/openstack/generate-yaml.sh (hunk 8 above)
Comment (gyliu513) [reply cmd/clusterctl/examples/openstack/generate-yaml.sh]: @hchenxa you should configure the `cacert` as the path to a CA Cert, but not the content of the cert.

```
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/vendor/github.com/gophercloud/utils/openstack/clientconfig/results.go#L33-L35
```

**Comment by jichenjc** [comment]
if there is no one to try , I can try to download to local and make some sniff test ... I got a https env now :)

File: cmd/clusterctl/examples/openstack/generate-yaml.sh (hunk 8 above)
Comment (hchenxa) [reply cmd/clusterctl/examples/openstack/generate-yaml.sh]: sure

**Comment by hchenxa** [comment]
@gyliu513 @jichenjc , I create create the VM with cacert now.

**Comment by gyliu513** [comment]
Thanks @hchenxa 

/lgtm

/cc @jichenjc @chrigl

File: cmd/clusterctl/examples/openstack/provider-component/user-data/ubuntu/templates/worker-user-data.sh [hunk 9]
```diff
@@ -79,6 +79,8 @@ CLUSTER_DNS_SERVER=$(prips ${SERVICE_CIDR} | head -n 11 | tail -n 1)
 
 # Write the cloud.conf so that the kubelet can use it.
 echo $OPENSTACK_CLOUD_PROVIDER_CONF | base64 -d > /etc/kubernetes/cloud.conf
+mkdir /etc/certs
+echo $OPENSTACK_CLOUD_CACERT_CONFIG | base64 -d > /etc/certs/cacert
```
Comment (jichenjc) [inline cmd/clusterctl/examples/openstack/provider-component/user-data/ubuntu/templates/worker-user-data.sh]: Just curious 
this is only for ubuntu, do we have a centos version? I didn't see it ...

File: cmd/clusterctl/examples/openstack/generate-yaml.sh [hunk 10]
```diff
@@ -186,14 +199,17 @@ if [[ "$PROVIDER_OS" == "coreos" ]]; then
 else
```
Comment (jichenjc) [inline cmd/clusterctl/examples/openstack/generate-yaml.sh]: you updated centos and ubuntu
but how about coreos ? I didn't see the change ,maybe a follow up??

**Comment by jichenjc** [comment]
thanks @hchenxa and @hidekazuna 
I think we may leave this for a few days to get some opinion then we can merge it 

in addition, I had some comments above, can you help to take a look @hidekazuna ? thank you

File: cmd/clusterctl/examples/openstack/provider-component/user-data/ubuntu/templates/worker-user-data.sh (hunk 9 above)
Comment (hidekazuna) [reply cmd/clusterctl/examples/openstack/provider-component/user-data/ubuntu/templates/worker-user-data.sh]: @jichenjc Thanks pointing out. I added in https://github.com/kubernetes-sigs/cluster-api-provider-openstack/commit/b06fd90d5eaff2c379e3322bf8bcbc5e3a60f3bc

File: cmd/clusterctl/examples/openstack/generate-yaml.sh (hunk 10 above)
Comment (hidekazuna) [reply cmd/clusterctl/examples/openstack/generate-yaml.sh]: @jichenjc Thanks commenting.i added code for coreos in https://github.com/kubernetes-sigs/cluster-api-provider-openstack/commit/b06fd90d5eaff2c379e3322bf8bcbc5e3a60f3bc

**Comment by gyliu513** [comment]
@hidekazuna `make generate_yaml_test` failed, can you help check?

```
mv kustomize_1.0.11_linux_amd64 /usr/local/bin/kustomize
chmod +x /usr/local/bin/kustomize
# Create a dummy file for test only
echo 'clouds' > dummy-clouds-test.yaml
cmd/clusterctl/examples/openstack/generate-yaml.sh -f dummy-clouds-test.yaml openstack ubuntu dummy-make-auto-test
Generating SSH key files for machine controller.
Generating public/private rsa key pair.
Created directory '/root/.ssh'.
Your identification has been saved in /root/.ssh/openstack_tmp.
Your public key has been saved in /root/.ssh/openstack_tmp.pub.
The key fingerprint is:
SHA256:L90G006kSqwbMxM0lT0drVo+4bXNMIGj7bsPZaG91j8 root@5807cc27-67fb-11e9-be50-0a580a6c05bf
The key's randomart image is:
+---[RSA 2048]----+
|        .o ..+   |
|       .. o + o  |
|      o    +.o.. |
|     . o  .+=o+. |
|      . S +*+o+* |
|       + +.*=o.oo|
|      * o o =oo .|
|       * . ..o E.|
|      .     .o. o|
+----[SHA256]-----+
cat: null: No such file or directory
Makefile:78: recipe for target 'generate_yaml_test' failed
make: *** [generate_yaml_test] Error 1
+ EXIT_VALUE=2
+ set +o xtrace
```

**Comment by hidekazuna** [comment]
> @hidekazuna `make generate_yaml_test` failed, can you help check?
> 
> ```
> mv kustomize_1.0.11_linux_amd64 /usr/local/bin/kustomize
> chmod +x /usr/local/bin/kustomize
> # Create a dummy file for test only
> echo 'clouds' > dummy-clouds-test.yaml
> cmd/clusterctl/examples/openstack/generate-yaml.sh -f dummy-clouds-test.yaml openstack ubuntu dummy-make-auto-test
> Generating SSH key files for machine controller.
> Generating public/private rsa key pair.
> Created directory '/root/.ssh'.
> Your identification has been saved in /root/.ssh/openstack_tmp.
> Your public key has been saved in /root/.ssh/openstack_tmp.pub.
> The key fingerprint is:
> SHA256:L90G006kSqwbMxM0lT0drVo+4bXNMIGj7bsPZaG91j8 root@5807cc27-67fb-11e9-be50-0a580a6c05bf
> The key's randomart image is:
> +---[RSA 2048]----+
> |        .o ..+   |
> |       .. o + o  |
> |      o    +.o.. |
> |     . o  .+=o+. |
> |      . S +*+o+* |
> |       + +.*=o.oo|
> |      * o o =oo .|
> |       * . ..o E.|
> |      .     .o. o|
> +----[SHA256]-----+
> cat: null: No such file or directory
> Makefile:78: recipe for target 'generate_yaml_test' failed
> make: *** [generate_yaml_test] Error 1
> + EXIT_VALUE=2
> + set +o xtrace
> ```

**Comment by jichenjc** [comment]
copy the link I added in the slack channel :)

@hidekazuna the https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/224 failed with a new test I added recently https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/322, at least it works when I submit the PR with my local stuffs and gate, please let me know whether some edge case triggered by your PR and any info needed by me ,thanks

File: cmd/clusterctl/examples/openstack/provider-component/user-data/ubuntu/templates/worker-user-data.sh (hunk 9 above)
Comment (hidekazuna) [reply cmd/clusterctl/examples/openstack/provider-component/user-data/ubuntu/templates/worker-user-data.sh]: Thanks added this for centos.

File: cmd/clusterctl/examples/openstack/generate-yaml.sh (hunk 10 above)
Comment (hidekazuna) [reply cmd/clusterctl/examples/openstack/generate-yaml.sh]: I Updated files for coreos.

**Comment by jichenjc** [comment]
I think we are ready for this one, as the @hchenxa already tested it

**Comment by gyliu513** [comment]
LGTM

Leave this to @chrigl for approve.
